data-prepper-core/src/main/java/org/opensearch/dataprepper/pipeline/Pipeline.java

```diff
@@ -211,7 +212,21 @@ public void execute() {
                 final SourceCoordinator sourceCoordinator = sourceCoordinatorFactory.provideSourceCoordinator(partionProgressModelClass, name);
                 ((UsesSourceCoordination) source).setSourceCoordinator(sourceCoordinator);
             }
-            source.start(buffer);
+
+            sinkExecutorService.submit(() -> {
+                long retryCount = 0;
+                while (!isReady() && !isStopRequested()) {
+                    if (++retryCount % 60 == 0) {
+                        LOG.info("Pipeline {} Waiting for Sink to be ready", name);
+                    }
+                    try {
+                        Thread.sleep(1000);
+                    } catch (Exception e){}
+                }
+                LOG.info("Pipeline {} Sink is ready, starting source...", name);
+                source.start(buffer);
+            }, null);
+
             LOG.info("Pipeline [{}] - Submitting request to initiate the pipeline processing", name);
```

Comment:
Let's include this in the async thread that initialize the source in. And we can keep the existing initialization order as was currently there.

It could also help to move this into an inner class rather than a lambda since there is quite a bit there.